### DISCUSSION TITLE
socket.io connect_error timeout on localhost

### DISCUSSION THREAD

**Discussion by tehnaz** [socketio/socket.io#4730] [Q&A]
Hello, 

We have a NodeJS server and ReactJS client communicating via socket.io.
This setup has been working for years. We are suddenly experiencing the following connect_error (without any changes in code):
![image](https://github.com/socketio/socket.io/assets/70631013/0194f33a-3882-401b-9fbd-30b309c2782e)

We are using transport 'polling' that is later upgraded to 'websocket' after the handshake, but in this case the polling requests go into pending state and then timeout.
The issue is only observed on some machines while on the others it is still working fine on localhost.
The same code is also working fine on the QA, Stage and Prod deployment environments.

Furthermore, we have a multi-node setup, therefore, we use the following packages to run our server on multiple CPU cores - cluster, @socket.io/sticky and socket.io-redis (now @socket.io/redis-adapter).
The React app runs on port 3000. The NodeJS app includes an HTTP server running on port 8090 and the websocket server running on port 8092.

Current versions:
OS: Windows 11
NodeJS: 14.16
socket.io: 3.1.2
@socket.io/sticky: 1.0.1
@socket.io/redis: 6.0.1
redis: 3.0.2

Since @socket.io/redis is now renamed to @socket.io/redis-adapter, we upgraded it from 6.0.1 to 8.2.1 and also migrated from socket.io 3.1.2 and 4.6.1. We have also tried upgrading the other related packages but we still see the same error.
Disabling the antivirus and firewall on the windows machine hasn't helped either.
However, we have observed that if we use transport 'websocket' instead of 'polling', the error goes away.

Based on Chrome network tab, network port monitoring and debug logs of socket.io, we have observed that when long polling starts, 1-2 polling requests are reaching the server, while the others are not.

Browser:
![image](https://github.com/socketio/socket.io/assets/70631013/2a870cff-98e0-47a7-8a4d-061ce70c6506)

Port traffic:
![image](https://github.com/socketio/socket.io/assets/70631013/79e5b714-af81-4869-b0dd-6c4669c6cf5d)

socket.io logs:
[logs.txt](https://github.com/socketio/socket.io/files/11623492/logs.txt)

Code Snippets:
Client:
```
import { io } from 'socket.io-client';
socket = io(payload,
  {
    transportOptions: {
      polling: {
        extraHeaders: {
          'Authorization': 'Bearer ${payload.token}'
        }
      }
    }
  });
socket.on('connect', () => {
...
});
socket.on('connect_error', () => {
...
});
socket.on('message', () => {
...
});
socket.on('disconnect', () => {
...
});
```

Server:
Master:
```
const cluster = require('cluster');
const http = require('http');
const { setupMaster } = require('@socket.io/sticky');

const httpServer = http.createServer();
setupMaster(httpServer, {
    loadBalancingMethod: 'least-connection',
});
httpServer.listen(8092);

cluster.setupMaster({
    ...
});

for (let i = 0; i < numberOfCores; i += 1) {
    cluster.fork();
}
cluster.on('exit', (worker) => {
    cluster.fork();
});
```

Worker:
```
const { setupWorker } = require('@socket.io/sticky');
const socketIo = require('socket.io');
const redis = require('redis');
const { createAdapter } = require('@socket.io/redis-adapter');

this.io = socketIo(http.createServer() {
    pingInterval,
    pingTimeout,
    cors: {
        origin: [https://localhost:3000],
        methods: ['GET', 'POST', 'OPTIONS'],
        allowedHeaders: ['Authorization'],
        credentials: true,
        preflightContinue: false,
        optionsSuccessStatus: 204,
    },
    cookie: {
        name: 'io',
        httpOnly: false,
    },
    allowEIO3: true,
});

setupWorker(this.io);

const redisClient = redis.createClient(credential);
const subClient = redisClient.duplicate();
const  adapter = createAdapter(redisClient, subClient);
this.io.adapter(adapter)
```

What could be the issue with polling + websocket?
Kindly assist.

Thanks
Tehnaz

**Comment by darrachequesne** [comment]
Hi! I see you are using HTTPS (`origin: [https://localhost:3000]`), could the error be linked to a self-signed certificate?

> we have observed that when long polling starts, 1-2 polling requests are reaching the server, while the others are not.
> The issue is only observed on some machines while on the others it is still working fine on localhost.

That sounds indeed like a firewall issue or something else getting in the way.

**Reply by tehnaz** [reply]
Hi.
Thanks for the reply.

HTTPS is only for the API server with a self-signed certificate. The socket server runs on HTTP locally.
I did turn off the Windows firewall and disabled the antivirus but that didn't help.

Another thing to note is that if I disable the multi-node setup, it works fine.

Anything else you think I should try? Any other external factors to consider?
Could it be related to Windows updates?

**Reply by tehnaz** [reply]
Hi @darrachequesne 

We tried running the API server also on HTTP instead of HTTPS and the socket timeout issue wasn't reproducible.
But we do need HTTPS for some other features of the app.

Few queries here:
1. How did making the API server HTTP help? Perhaps because the UI client is also running on HTTP now? So does this mean that an HTTPS socket client cannot connect to an HTTP socket server? 
2. This same setup was working since a long time, how could it stop suddenly, that too only on some machines?
3. Also considering the fact that the issue is not reproducible when using the websocket transport and disabing the mutli-node setup, could the issue be related to socket.io/sticky?
4. The connect_error 'timeout' doesn't explain much. Any other ways to debug this further?

Thanks.

**Reply by darrachequesne** [reply]
> Also considering the fact that the issue is not reproducible when using the websocket transport and disabing the mutli-node setup, could the issue be related to socket.io/sticky?

`@socket.io/sticky` is not currently compatible with HTTPS (see [here](https://github.com/socketio/socket.io-sticky/issues/3)), maybe that explains your issue?

> This same setup was working since a long time, how could it stop suddenly, that too only on some machines?

This is weird though.

**Reply by tehnaz** [reply]
Hi @darrachequesne 

We were finally able to understand the cause of the issue.
We were using socket.io/sticky version 1.0.1 but due to a mistake, our package-lock.json had version 1.0.3
The developers who didn't update their node modules after this change were able to connect the sockets while the others encountered the issue.
Therefore, we are now able to establish the socket connection successfully using socket.io/sticky 1.0.1 with socket.io v3 and v4 both.

However, while dedugging the same, we found out that our client side code that uses extraHeaders is not working with socket.io/sticky 1.0.3 and that was causing the issue:

That is, this doesn't work:
```
socket = io(payload,
  {
    transportOptions: {
      polling: {
        extraHeaders: {
          'Authorization': 'Bearer ${payload.token}'
        }
      }
    }
  });
```
but this does:
```
{
    transports: ['polling', 'websocket'],
    auth: { token: `${payload.token}` },
 }
```

We also tried with socket.io/sticky 1.0.2 and observed the same results as 1.0.3.
Only 1.0.1 is working properly.
I checked the release notes and couldn't find anything that could break this.

Why is transportOptions with extraHeaders not working with socket.io/sticky 1.0.2 and 1.0.3?
Kindly assist. Thanks.

Regards,
Tehnaz

**Reply by darrachequesne** [reply]
Thanks for digging into this :+1:

[This commit](https://github.com/socketio/socket.io-sticky/commit/a124d0beb5c1b78be5b75f10153859a9e4672862) might be at fault then (included in `@socket.io/sticky@1.0.2`), but the CI only fails for Windows + Node.js 10: https://github.com/socketio/socket.io-sticky/actions/runs/5207695992

Regarding the `transportOptions` value, I don't know what could go wrong, maybe your CORS configuration is not valid? Do you use `cors.allowedHeaders`?

Reference: https://socket.io/docs/v4/server-options/#cors

**Reply by tehnaz** [reply]
The CI test result seems similar to our usecase but we are on Node 14.

Yes, we do use allowedHeaders:

```
this.io = socketIo(http.createServer() {
    pingInterval,
    pingTimeout,
    cors: {
        origin: [https://localhost:3000],
        methods: ['GET', 'POST', 'OPTIONS'],
        allowedHeaders: ['Authorization'],
        credentials: true,
        preflightContinue: false,
        optionsSuccessStatus: 204,
    },
    cookie: {
        name: 'io',
        httpOnly: false,
    },
    allowEIO3: true,
});
```

**Reply by tehnaz** [reply]
Hi @darrachequesne 

Do you see any issues with the above config?
Can we confirm this is a bug, should I report it on the socket.io-sticky page?

Regards,
Tehnaz

**Reply by darrachequesne** [reply]
OK, I was able to reproduce, the error indeed seems to come from the sticky package. I'll keep you updated.
